Sources/EdgeNetworkHandlers/CompletionHandlersManager.swift

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright 2020 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may obtain a copy
+// of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import AEPCore
+import AEPServices
+import Foundation
+
+/// When registering a completion handler for a request event, all the event handles will be returned at once when the entire streamed response was processed by the AEP Edge extension.
+/// This class uses a `ThreadSafeDictionary` for the internal mapping.
+class CompletionHandlersManager {
+    private let TAG = "CompletionHandlersManager"
+    private var completionHandlers =
+        ThreadSafeDictionary<String, (([EdgeEventHandle]) -> Void)>(identifier: "com.adobe.edge.completionHandlers")
+
+    // edge response handles for a event request id (key)
+    private var edgeEventHandles =
+        ThreadSafeDictionary<String, [EdgeEventHandle]>(identifier: "com.adobe.edge.edgeHandlesList")
+
+    static let shared = CompletionHandlersManager()
+
+    /// Registers a completion handler for the specified `requestEventId`. This handler is invoked when the Edge response content has been
+    /// handled entirely by the Edge extension, containing a list of `EdgeEventHandle`(s). This list can be empty or can contain one or multiple items
+    /// based on the request and the server side response.
+    ///
+    /// - Parameters:
+    ///   - forRequestEventId: unique event identifier for which the completion handler is registered; should not be empty
+    ///   - completionHandler: the completion handler that needs to be registered, should not be nil
+    func registerCompletionHandler(forRequestEventId: String, completion: (([EdgeEventHandle]) -> Void)?) {
```

Comment:
Is it possible to potentially register a completion handler twice for a single request id? I don't think there's a use case for this but I wonder if we should be safe and check if an existing completion handler is already mapped to the request id.

Reply:
Since this API is internal and used when sendEvent is called and a new Event is created, I think it is safe to say that there will not be multiple completion handlers associated with the same request id.